analyzer/src/main/kotlin/managers/Npm.kt

```diff
@@ -389,67 +371,87 @@ open class Npm(
-                    dependencyName, packages, newDependencyBranch
-                )?.let { dependencies += it }
+                    moduleDir = dependencyModuleDir,
+                    scopes = setOf("dependencies", "optionalDependencies"),
+                    ancestorModuleDirs = dependencyModuleDirPath.subList(1, dependencyModuleDirPath.size),
+                    ancestorModuleIds = ancestorModuleIds + moduleId,
+                    packageType = "NPM"
+                )?.let { dependencies.add(it) }
+
+                return@forEach
             }
 
-            val packageInfo = packages[identifier]
-                ?: throw IOException("Could not find package info for $identifier")
-            return packageInfo.toReference(dependencies = dependencies)
-        } else if (rootModulesDir == startModulesDir) {
-            return getPackageReferenceForMissingModule(moduleName, rootModulesDir)
-        } else {
-            // Skip the package name directory when going up.
-            var parentModulesDir = startModulesDir.parentFile.parentFile
+            log.debug { "Could not find module dir for '$dependencyName' within: '${pathToRoot.joinToString()}'." }
+            getPackageReferenceForMissingModule(dependencyName, pathToRoot.first())
+        }
 
-            // For scoped packages we need to go one more directory up.
-            if (parentModulesDir.name.startsWith("@")) {
-                parentModulesDir = parentModulesDir.parentFile
-            }
+        return PackageReference(id = moduleId, dependencies = dependencies)
+    }
 
-            // E.g. when using Yarn workspaces, the dependencies of the projects are consolidated in a single top-level
-            // "node_modules" directory for de-duplication, so go up.
-            log.debug {
-                "Could not find package file for '$moduleName' in '$startModulesDir', looking in '$parentModulesDir' " +
-                        "instead."
-            }
+    private data class ModuleInfo(
+        val name: String,
+        val version: String,
+        val dependencyNames: Set<String>
+    )
+
+    private fun getModuleInfo(moduleDir: File, scopes: Set<String>): ModuleInfo {
+        val packageJsonFile = moduleDir.resolve("package.json")
+        val json = jsonMapper.readTree(packageJsonFile)
+
+        return ModuleInfo(
+            name = json["name"].textValue(),
+            version = json["version"].textValue(),
+            dependencyNames = scopes.map { scope ->
+                json[scope]?.fieldNames()?.asSequence()?.toSet().orEmpty()
+            }.flatten().toSet()
+        )
+    }
 
-            return buildTree(rootModulesDir, parentModulesDir, moduleName, packages, dependencyBranch)
+    private fun findDependencyModuleDir(dependencyName: String, searchModuleDirs: List<File>): List<File> {
+        searchModuleDirs.forEachIndexed { index, moduleDir ->
+            val dependencyModuleDir = moduleDir.resolve("node_modules/$dependencyName")
```

Comment:
Does this work because the `dependencyName` for namespaced dependencies contains a slash, like "@namespace/depednency"? If so it's worth a code comment IMO.

Reply:
Done.